README.md

```diff
@@ -0,0 +1,303 @@
+DataObjectEventListener::create(
+    function (DataObjectEvent $event) {
+        // Handle create/update operations
+    },
+    [Member::class, Group::class],
+    [Operation::CREATE, Operation::UPDATE]
+)->selfRegister();
+```
+
+### Available Operations
+
+The following operations are automatically tracked:
+
+- `Operation::CREATE` - When a DataObject is first written
+- `Operation::UPDATE` - When an existing DataObject is modified
+- `Operation::DELETE` - When a DataObject is deleted
+- `Operation::PUBLISH` - When a versioned DataObject is published
+- `Operation::UNPUBLISH` - When a versioned DataObject is unpublished
+- `Operation::ARCHIVE` - When a versioned DataObject is archived
+- `Operation::RESTORE` - When a versioned DataObject is restored from archive
+
+### Accessing Event Data
+
+The `DataObjectEvent` class provides several methods to access information about the event:
+
+```php
+DataObjectEventListener::create(
+    function (DataObjectEvent $event) {
+        $object = $event->getObject();        // Get the affected DataObject
+        $class = $event->getObjectClass();    // Get the class name
+        $operation = $event->getOperation();  // Get the operation type
+        $version = $event->getVersion();      // Get version number (if versioned)
+        $member = $event->getMember();        // Get the Member who performed the action
+        $time = $event->getTimestamp();       // Get when the event occurred
+    },
+    [DataObject::class]
+)->selfRegister();
+```
+
+`DataObjectEvent` is configured to be serializable so it can easily be stored for later use.
+
+Note that `DataObjectEvent` doesn't store the actual DataObject instance that caused the event to be fired. 
+`DataObjectEvent::getObject()` will refetch the latest version of the DataObject from the database ... which will 
+return `null` if the DataObject has been deleted.
+
+`DataObjectEvent::getObject(true) will attempt to retrieve the exact version of the DataObject that fired the event,
+assuming it was versioned.
+
+## Testing Your Events
+
+### Writing Event Tests
+
+When testing your event listeners, you'll need to:
+1. Dispatch your events
+2. Run the event loop
+3. Assert the expected outcomes
+
+Here's an example test:
+
+```php
+use Revolt\EventLoop;
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Core\Injector\Injector;
+use ArchiPro\Silverstripe\EventDispatcher\Service\EventService;
+
+class MyEventTest extends SapphireTest
+{
+    public function testMyCustomEvent(): void
+    {
+        // Create your test event
+        $event = new MyCustomEvent('test message');
+        
+        // Get the event service
+        $service = Injector::inst()->get(EventService::class);
+        
+        // Add your test listener ... or if you have already
+        $wasCalled = false;
+        $service->addListener(
+            MyCustomEvent::class, 
+            [MyCustomEventListener::class, 'handleEvent']
```

Comment:
If we have interfaces, can we simplify the API by just passing in the listener class? Might be nice to be able to pass in custom handler method names but.. :shrug: